### ISSUE TITLE
runway tfenv failing

### ISSUE THREAD

**Issue by shobhit063** [onicagroup/runway#1394]
### Discussed in https://github.com/onicagroup/runway/discussions/1384

<div type='discussions-op-text'>

<sup>Originally posted by **pataraco** April 15, 2022</sup>
Hello,

This started happening yesterday Apr 13, 2022

We've tried all sort of combinations of runway/terraform/python versions and cannot seem to make this work anymore.

This is affect our CI/CD pipeline which can no longer install the version of terraform specified in the `runway.yaml` file because `runway`, i believe, is using `runway tfenv` to install the desired version of terraform

Any help would be greatly appreciated. Until then i'll be creating my own "wrapper" to install the desired versions of terraform myself into the directory runway is looking for `$HOME/.tfenv/versions`

Thanks!

```
[sandbox] 😬-bash-⌲ runway tfenv run --version
Traceback (most recent call last):
  File "/Users/patrick.raco/.local/bin/runway", line 8, in <module>
    sys.exit(cli())
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/_cli/main.py", line 31, in invoke
    return super(_CliGroup, self).invoke(ctx)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/_cli/commands/_tfenv/_run.py", line 36, in run
    ctx.exit(subprocess.call([TFEnvManager().install()] + list(args)))
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/env_mgr/tfenv.py", line 298, in install
    for i in get_available_tf_versions(include_prerelease_versions)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/env_mgr/tfenv.py", line 94, in get_available_tf_versions
    reverse=True,
  File "/Users/patrick.raco/.pyenv/versions/3.7.12/lib/python3.7/distutils/version.py", line 52, in __lt__
    c = self._cmp(other)
  File "/Users/patrick.raco/.pyenv/versions/3.7.12/lib/python3.7/distutils/version.py", line 337, in _cmp
    if self.version < other.version:
TypeError: '<' not supported between instances of 'str' and 'int'
```</div>

**Comment by shobhit063** [comment]
@pataraco did you get any work around for this issue , as we still facing same issue . and it delays our infrastructure promotion

https://github.com/onica please help here

11:59:00 [63] Failed to execute script 'runway-script' due to unhandled exception!
11:59:00 Traceback (most recent call last):
11:59:00 File "runway/module/terraform.py", line 550, in run
11:59:00 File "runway/module/terraform.py", line 341, in handle_parameters
11:59:00 File "backports/cached_property/init.py", line 80, in get
11:59:00 File "runway/module/terraform.py", line 132, in auto_tfvars
11:59:00 File "backports/cached_property/init.py", line 80, in get
11:59:00 File "runway/module/terraform.py", line 209, in version
11:59:00 File "backports/cached_property/init.py", line 80, in get
11:59:00 File "runway/env_mgr/tfenv.py", line 307, in version
11:59:00 File "runway/env_mgr/tfenv.py", line 119, in get_available_tf_versions
11:59:00 File "distutils/version.py", line 52, in lt
11:59:00 File "distutils/version.py", line 337, in _cmp
11:59:00 TypeError: '<' not supported between instances of 'str' and 'int'

**Comment by pataraco** [comment]
> @pataraco did you get any work around for this issue , as we still facing same issue . and it delays our infrastructure promotion
> 
> https://github.com/onica please help here
> 
> 11:59:00 [63] Failed to execute script 'runway-script' due to unhandled exception! 11:59:00 Traceback (most recent call last): 11:59:00 File "runway/module/terraform.py", line 550, in run 11:59:00 File "runway/module/terraform.py", line 341, in handle_parameters 11:59:00 File "backports/cached_property/init.py", line 80, in get 11:59:00 File "runway/module/terraform.py", line 132, in auto_tfvars 11:59:00 File "backports/cached_property/init.py", line 80, in get 11:59:00 File "runway/module/terraform.py", line 209, in version 11:59:00 File "backports/cached_property/init.py", line 80, in get 11:59:00 File "runway/env_mgr/tfenv.py", line 307, in version 11:59:00 File "runway/env_mgr/tfenv.py", line 119, in get_available_tf_versions 11:59:00 File "distutils/version.py", line 52, in lt 11:59:00 File "distutils/version.py", line 337, in _cmp 11:59:00 TypeError: '<' not supported between instances of 'str' and 'int'

yeah, i wrote a couple of bash scripts to pre-install terraform for runway

**Comment by pataraco** [comment]
@shobhit063 

here is a quick workaround we've implemented while waiting for the new patched versions of runway

### install-required-terraform-versions.sh
```
#!/bin/bash
#
# purpose: install required terraform versions listed in
#          either or both $TERRAFORM_VERSION_FILE and/or runway.yaml files
#

# enable bash debugging if set on the pipeline
if grep -Fiwq "true" <<< "$SYSTEM_DEBUG"; then set -x; fi

# globals
# note: DEVOPS_ADO_SCRIPTS defined in terraform/common/azure-variables.yaml
DEVOPS_SCRIPTS=$DEVOPS_ADO_SCRIPTS
TERRAFORM_VERSION_FILE=".terraform-version"

function get_tf_versions {
  local _runway_file=$1
  awk '$1~/terraform_version/ {tfvfound="true"; next}; (tfvfound=="true") {print $2; tfvfound="false"}' "$_runway_file" | sort -u
}

# grab the version in the TERRAFORM_VERSION_FILE if it exists
if [[ -f "$TERRAFORM_VERSION_FILE" ]]; then
  required_terraform_version=$(cat "$TERRAFORM_VERSION_FILE")
  if [[ -n "$required_terraform_version" ]]; then
    echo "found this terraform version specified in the '$TERRAFORM_VERSION_FILE' file: $required_terraform_version"
  else
    echo "Note: didn't find a required terraform version specified in a '$TERRAFORM_VERSION_FILE' file"
  fi
else
  echo "Note: file '$TERRAFORM_VERSION_FILE' not found - not obtaining a required terraform version from there"
fi

# find versions specified in the runway.yml or runway.yaml file
for runway_file in runway.yml runway.yaml; do
  if [[ -f "$runway_file" ]]; then
    required_terraform_versions_in_runway_yaml=$(get_tf_versions "$runway_file")
  fi
done
if [[ -n "$required_terraform_versions_in_runway_yaml" ]]; then
  # shellcheck disable=SC2086
  echo "found terraform version(s) specified in the 'runway' YAML file:" $required_terraform_versions_in_runway_yaml
else
  echo "Note: didn't find any terraform versions specified in the 'runway' YAML file"
fi

# concatenate all found versions and install them
# shellcheck disable=SC2028
required_terraform_versions=$(echo -e "$required_terraform_version\n$required_terraform_versions_in_runway_yaml" | sort -u)
# loop through list of required terraform versions and install them
if [[ -n "$required_terraform_versions" ]]; then
  # shellcheck disable=SC2086
  echo "full list of terraform versions to be installed:" $required_terraform_versions
  for terraform_version in $required_terraform_versions; do
    echo "Installing Terraform version: $terraform_version"
    "$DEVOPS_SCRIPTS/install-terraform.sh" "$terraform_version"
  done
else
  echo "Error: didn't find any terraform versions to install"
  echo "Cannot pre-install Terraform versions required"
  exit 1
fi
```

**Comment by pataraco** [comment]
@shobhit063 

and here

### install-terraform.sh

```
#!/bin/bash
#
# purpose: install terraform version
#
# description: simple script to install a specific terraform version
# 
# note: the versions of terraform are installed into the
#       directory that runway is looking for so that
#       runway does not try to install it itself using `runway tfenv`
#       which is currently broken as of (13-APR-2022)
#

# enable bash debugging if set on the pipeline
if grep -Fiwq "true" <<< "$SYSTEM_DEBUG"; then set -x; fi

THIS_SCRIPT=$(basename "$0")
TF_ENV_DIR="$HOME/.tfenv/versions"  # this is where runway is looking
TF_RELEASES_URL="https://releases.hashicorp.com"
USAGE=\
"usage:
  $THIS_SCRIPT VERSION      install specified terraform version
  $THIS_SCRIPT -s | --show  show installed terraform versions
  $THIS_SCRIPT -h | --help  disploy this help/usage"

# create the directory if it doesn't exist
[[ ! -d "$TF_ENV_DIR" ]] && mkdir -p "$TF_ENV_DIR"

# get the version to install
tf_version_desired=$1

function show_installed_versions {
  # shellcheck disable=SC2086
  if ls -d $TF_ENV_DIR/* &> /dev/null; then
    echo "these are the installed versions:"
    ls "$TF_ENV_DIR"
  else
    echo "there are no installed versions"
  fi
}

# command: help - show help
if [[ "$tf_version_desired" == "-h" ]] || [[ "$tf_version_desired" == "--help" ]]; then
  echo "this function helps to install a specific terraform version"
  echo "$USAGE"
# command: show - show terraform versions installed/available
elif [[ "$tf_version_desired" == "-s" ]] || [[ "$tf_version_desired" == "--show" ]]; then
  show_installed_versions
# command: "install" (default) - install the specified terraform version
elif [[ -n "$tf_version_desired" ]]; then
  if [[ -x "$TF_ENV_DIR/$tf_version_desired/terraform" ]]; then
    echo "desired version ($tf_version_desired) already installed in '$TF_ENV_DIR'"
  else
    echo "did not find desired version ($tf_version_desired) in '$TF_ENV_DIR'"
    show_installed_versions
    echo "going to attempt to install it"
    host_arch=$(uname -m)
    case $host_arch in
      x86_64) arch=amd64;;
      arm64) arch=arm64;;
      *) arch=unknown;;
    esac
    host_os=$(uname -s)
    case $host_os in
      Darwin) os=darwin;;
      Linux) os=linux;;
      *) os=unknown;;
    esac
    if [[ "$arch" == "unknown" ]] && [[ "$os" == "unknown" ]]; then
      echo "cannot determine correct terraform binary to download for (Arch: $host_arch, O.S.: $host_os)"
      exit
    fi
    tf_version_zip_name="terraform_${tf_version_desired}_${os}_${arch}.zip"
    tf_version_url="${TF_RELEASES_URL}/terraform/${tf_version_desired}/${tf_version_zip_name}"
    if curl -s -f -o "/tmp/${tf_version_zip_name}" "${tf_version_url}"; then
      [[ ! -d "$TF_ENV_DIR" ]] && mkdir "$TF_ENV_DIR"
      unzip "/tmp/${tf_version_zip_name}" -d "$TF_ENV_DIR/$tf_version_desired"
      echo "successfully installed desired version ($tf_version_desired) in '$TF_ENV_DIR'"
      show_installed_versions
    else
      tf_available_versions="$(curl -s "${TF_RELEASES_URL}/terraform/" | grep -oE 'terraform_\d+\.\d+.\d+' | sort -uV | sed 's/^terraform_//' | tr '\n' ',' | sed 's/,/, /g')"
      echo "cannot find desired version ($tf_version_desired) at '$TF_RELEASES_URL'"
      echo "these are the available versions:"
      echo "  $tf_available_versions"
      exit
     fi
  fi
# command: none - show usage
else
  echo "error: incorrect usage"
  echo "$USAGE"
fi
```

**Comment by shobhit063** [comment]
@pataraco Thank you so much , your solution worked in my enviroment

**Comment by mbordash2** [comment]
@shobhit063 This issue should be resolved in `v2.6.2`; please let us know.
